### DISCUSSION TITLE
Review comments not shown in "Files changed" if a commit affects the comment's associated line

### DISCUSSION THREAD

**Discussion by renozao** [community/community#23138] [New to GitHub]
<p>Consider the following situation:</p>
<ol>
<li>Create a pull request</li>
<li>Insert comments at some individual lines</li>
<li>Submit review =&gt; One can see all the comments embedded within the files, right after their respective associated lines</li>
<li>Make and commit a change to the line associated with one of the comments, referencing the pull request in the commit message</li>
</ol>
<p>Then, going to the “Files changed” tab of the commit, one can see the changes, as well as review comments, but the review comment that was associated with the line that has been modified is <strong>not</strong> shown.</p>
<p>Having the comment thread shown there would enable resolving the conversation straightway, since the commit would typically address the comment.</p>
<p>Is this an intended behaviour?</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>Yes, this is the intended behavior. If the line that the comment was in relation to changed, the automatic tools have no way of knowing if the comment targeted at the old line still makes sense on the new line. You can, however, always see the full conversation on the Conversation tab. Any comments referencing lines that have been changed show with an “Outdated” flag:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/3/3d28e281d58a44db96bdad8b2955ffbeff6f9e5a.png" data-download-href="/uploads/short-url/8J2IHnQaWWnOjdCkeYdVTNq9s5Y.png?dl=1" title="Screen Shot 2019-03-29 at 7.25.59 AM.png"><img src="https://user-images.githubusercontent.com/1038121/181075068-824d4720-c9ab-4c95-98d7-4d0744015bed.png" alt="Screen Shot 2019-03-29 at 7.25.59 AM.png" data-base62-sha1="8J2IHnQaWWnOjdCkeYdVTNq9s5Y" width="690" height="381" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/3/3d28e281d58a44db96bdad8b2955ffbeff6f9e5a_2_690x381.png, https://aws1.discourse-cdn.com/github/optimized/2X/3/3d28e281d58a44db96bdad8b2955ffbeff6f9e5a_2_1035x571.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/3/3d28e281d58a44db96bdad8b2955ffbeff6f9e5a_2_1380x762.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/3d28e281d58a44db96bdad8b2955ffbeff6f9e5a_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2019-03-29 at 7.25.59 AM.png</span><span class="informations">1954×1080 223 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>You can see this example for yourself <a href="https://github.com/lee-dohm/test-repo/pull/6">in my test repository</a>.</p>
<p>I hope that helps!</p>

**Reply by AnnaMadsenKnowit** [reply]
> the automatic tools have no way of knowing if the comment targeted at the old line still makes sense on the new line.

That is the exact point of still showing the comment. If it no longer makes sense, most likely because the addressed issue has been fixed, then the comment can be resolved. The comparison between the comment and and the updated code our most important tool in resolving comments. What does not make sense is that GitHub does not support this.

**Reply by dadaday** [reply]
+1

**Reply by vedro-compota** [reply]
@AnnaMadsenKnowit  +1 agree with you

**Reply by stephenjason89** [reply]
@lee-dohm is there any news on this after 4 years?
Would be nice if the outdated comment atleast shows how the code looks with the new changes.

**Reply by jwosty** [reply]
I've become absolutely paranoid of missing comments (as a reviewer) over time, and I'm now realizing that this is why. It makes it feel unstable - like conversation threads can seemingly disappear but still keep the PR open. It makes the tool feel flaky and unreliable.

I really just want it to show me *all* unresolved conversations, even potentially outdated ones. Not doing so makes it systematically difficult to determine whether or not a comment was truly addressed by a change intended to do so. Often times it's just a one-line change that makes the conversation vanish from the Files Changed view! It's maddening.

I'm hitting my thumb with this hammer so often that for the first time ever, I'm considering finding a different review tool for myself and my team. I honestly wouldn't have considered using something else if it weren't for this one bug / missing feature. Add this, and GH code review will be a pretty decent tool.

**Reply by hsearcy** [reply]
I just stumbled upon this after doing a self-review of some of my code and then cleaning up a few lines, then seeing all of my comments that I left for my team were now marked as Outdated and not visible in the review page... It's madness that there is no way to at least manually show Outdated comments or mark them as not Outdated. Seriously, how much effort does it take to make the `Outdated` label removable?

**Reply by jen-chitty-se** [reply]
Today I found myself advocating to move away from GitHub to something else because of this very issue. I happen to know for a fact that Gerrit handles this well. I find it extremely surprising that anyone would think the way that GitHub handles this is acceptable, let alone preferable.

Unfortunately, I suspect that this conversation is being ignored by GitHub staff, because this issue has been marked "Answered." 😢

**Reply by Caitlin-Syntax** [reply]
Just chiming in to bump the thread (if it does, in fact, get bumped anywhere). This is supremely annoying behavior. 

The only standalone workaround is to keep two tabs open to bounce back and forth between the Conversation thread and the Files Changed screen, and even then you can't easily see exactly what's going on. I also didn't realize at first that the default behavior is not to "show all changes" in the commit dropdown on the Files Changed screen. I couldn't figure out why I wasn't seeing my coworker's latest commits addressing my comments, which is literally the whole point of reviewing a PR. (Another workaround that at least sometimes works is to leave comments only at the file level with a mention of the line number, rather than inline with the code, but that's annoying for both reviewer and developer, especially on longer files.)

And if this was the only issue, maybe I wouldn't be so thoroughly vexed, but there are too many UI/UX issues that add up to a massive pain in my hindquarters every time I need to review a PR, which is nearly daily. It feels like no user testing in a real business environment has been done at all.

**Reply by tuffylock** [reply]
extremely frustrating to change something in a line that's unrelated to a discussion happening in the context of that line and then have people who are reviewing the code from the code view entirely miss that context/discussion. 

plenty of options to fall back on if you can't "be sure" its the same line! leave it there but put a note that context might have changed! move it to the top or bottom of the file! 

this behavior also encourages people to use comments that aren't inline, which is awful due to the entirely separate github issue of _being unable to reply to top level comments directly_

**Reply by ava351** [reply]
No, this is likely not intended behavior. Review comments should ideally persist in the "Files changed" tab after a commit that addresses them. This helps maintain conversation context. It might be a bug in the specific platform you're using.

**Reply by jchesshirAspire** [reply]
@ava351 Be sure to read my post about the links to other unanswered conversations.

**Reply by ava351** [reply]
alright

**Reply by bdaz** [reply]
@lee-dohm Regarding 

> "the automatic tools have no way of knowing if the comment targeted at the old line still makes sense on the new line."

This is a solvable problem—see how Phabricator (Differential) does it. The principles are: 
1. The changeset we want to view is _from the time the review was done_ → _to the latest changes_. When viewing that, review comments show up on the left when in split mode.
2. If you can't perfectly map from the context of a review comment to the latest changes, then either: (a) make a best guess, or failing that, (b) put the unmappable change at the top, with no corresponding code on the right side.

Most of the complexity of fixing this huge UX issue with GitHub is in implementing 2(a). But an implementation that only did 2(b) would be a massive upgrade!

**Reply by serhii-surazhskyi-gl** [reply]
The problem is still there.

It is frustrating and absolutely makes no sense in the normal review system to hide comments, regardless of their status.
Of course, if developers were using their review tool that would make obvious to them (yup, passive aggressive based on shithub quality)
Kindly check Gerriit implementation, how such situation is handled, even there are proposals from other developers in the thread.

"the automatic tools have no way of knowing if the comment targeted at the old line still makes sense on the new line." - that is not responsibility of the review system. It shall provide convenient way of viewing changes, commenting them and tacking review comments, it it is clear if they are fixed etc.

Why would so many people created so many comments about the issue if that was not important?

**Reply by Flaming-Dorito** [reply]
Why can't the comments still be shown in the files changed tab with the outdated tag so users know the line was changed?

Shameful that all other git repo managers ive used have never had this problem. Hell even Azure devops does a better job dealing with this and github cant figure it out.


If not by default there should atleast be a way to toggle this feature on?



Why do I need to have 2 tabs open with conversation and files changed to make sure im not missing comments when checking if I addressed them all while updating my PR?

**Comment by ghost** [comment]
<p><a class="mention" href="/lee-dohm">@lee-dohm</a> </p>
<p>This is a stupid and limiting design IMO. If I make a comment on a line, and that line is changed to fix what I commented about, I want to be able to verify that the comment is fixed.</p>
<p>I can’t do this in the Conversation tab view, since the fixed code isn’t shown there.</p>
<p>I can’t do this in the Files changed tab view (selecting the commit with the change), since the comment isn’t shown there.</p>
<p>Basically, this design forces me to open two different browser windows, and cross-comparing the Conversation view with the Files changed view.</p>
<p>Sure, the code file that the comment is in might have changed drastically, making the comment obsolete, but I rather think that the risk for that is very low compared to the chance of the file changing due to the actual comment.</p>

**Reply by Wim-Veen** [reply]
Exactly that. If I write a comment in a PR, I want to see how that comment is answered (either by reply, or by code change). If I have to look through all code changes to see what possibly has been done, it takes me ages to see if I can set the comment to resolved.

**Reply by jwosty** [reply]
What's especially frustrating is that even a one-line change causes this problem. For example:

1. Reviewer recommends a trivial change on line 100
2. Developer makes some simple one-line change on line 100 to address it

Now, the conversation is completely absent when looking at `Files Changed > Changes from all commits`. If the reviewer tries to use `Files Changed > Conversations` to locate the conversation, the comment shows up, but without the new code... Hence, the only way to verify if that newest *one line change* addresses the feedback, is to open two browser windows and line them up manually.

This get totally ridiculous to deal with in reviews with many trivial change requests.

**Comment by wlannistamps** [comment]
<p>Unfortunately, whether or not the line changed isn’t as much of a big deal as having to pop open a new tab to review the conversations at the same time as being able to view the updated code. It is actually harder for me to track the conversation to the changed line than it would be if the comment were still attached to an incorrect line. We used to deal with that all the time in Swarm; when it was off by a few lines, it was ok because we knew from the comment what the original code was, and could scroll back to wherever the new relevant line was. But it was much easier to immediately review the new code to the comment than it is to have to pop open two tabs, and then manually try to line up the new code to the comment.</p>
<p>Note to others, you can almost kind of sort of simulate this if you expand the commit drop down (upper left of the code panel) to include more than the past one commit. Unfortunately, you may accidentally then display code from other commits that occurred in between. But in this way I was actually able to see the original comment, and see the new code. I did this by resetting the view to split mode. I’m not sure this always works though, or works as intended.</p>

**Comment by aarontp** [comment]
<blockquote>
<p><a class="mention" href="/lee-dohm">@lee-dohm</a> wrote:</p>
</blockquote>
<blockquote>
<p>Yes, this is the intended behavior. If the line that the comment was in relation to changed, the automatic tools have no way of knowing if the comment targeted at the old line still makes sense on the new line. </p>
</blockquote>
<p>I understand that the tooling will not know if the comment makes sense at the new line, but it could be attached to the same line in the left pane, and then it can be left up to the user whether to resolve it or not.  This would make it *so* much easier to review large pull requests with lots of comments.</p>
<p>In the worst case, I would prefer to have the old comment still show up, but attached to the potentially wrong line if things have changed substantially.  At least it would be near the correct code, and it wouldn’t require me to flip back and forth between two windows, and manually look up where in the code the comment applies.</p>
<p>Can we file a feature request somewhere? :slight_smile:</p>

**Comment by zeptonaut** [comment]
<p>Totally agree. This makes it way more difficult than necessary to figure out if the comments that were marked “resolved” have in fact been resolved.</p>

**Comment by taurus227** [comment]
<p>I agree! This is the feature I would like to have: if I click the [Outdated] tag, I get a split view where on the left I see the commit I have commented on together with the comment and on the rigth side I see the newest code with all the changes that have happened since.</p>

**Reply by Caitlin-Syntax** [reply]
Bumping this response, four years later... This is exactly what I would want.

**Comment by brianhhq** [comment]
<p>of course we need this feature.</p>

**Comment by gpascale** [comment]
<p>Adding my voice to those who argue the current behavior is problematic. It’s entirely confusing that some comments will still show in the “files changed” view while others don’t. For a while, I didn’t even realize this behavior was happening and I’ve been missing a lot of these disappearing comments as a result.</p>
<p>As others have said, it’s really not a big deal if a comment doesn’t “make sense” anymore due to a change in the file. I can easily resolve it or create a new comment. It IS a big deal if I lose track of and fail to address important comments in a PR because they are missing. Jumping back and forth between two tabs is not a solution. The files changed tab is the sensible place to do the work of reviewing a PR as that is where the context is.</p>
<p>EDIT: Even worse, I’ve just realized that the “outdated” comments that show in the Conversation tab will only show the outdated code. Thus, there’s no easy way for me to verify that the change has addressed the feedback in order to resolve the comment. AFAICT, the only way is to switch to the files changed tab and scroll around looking for the change in question, verify the comment is addressed, then go back to Conversation view and mark as resolved. This is totally impractical.</p>

**Comment by lordzsolt** [comment]
<p>This is one of the stupidest “intended behaviors” ever. You need to have 2 tabs open to review any change because “Github attaches comments to commit hashes” (probably because that was the easiest way to implement it). I want to see <em>how</em> a comment got fixed.</p>
<p>Just look at Bitbucket for example. Even if lines get updates, comments still show up at the correct place. <strong>EVEN if you force-push</strong>, the comment remains.</p>

**Comment by lmussier** [comment]
<p>Is there any issue related to that topic we can raise ?<br>
I’m completely agree with above comments, this feature is useless like this.</p>

**Comment by kunickiaj** [comment]
<p>Yeah, this is one of the biggest gripes our users have that prevents us from migrating away from Gerrit.</p>

**Comment by Januszpl** [comment]
<p><a class="mention" href="/lee-dohm">@lee-dohm</a> I also have similar experience like other. I would like to see on the Conversation tab also changed code, not only “outdated” to be sure it was fixed properly. For larger PR it is problematic not having easy way to resolve issues and be sure they were fixed in proper way.</p>

**Comment by kostacasa** [comment]
<p>Another commenter mentioned this, but BitBucket gets it <em>much better</em> than the current GitHub behavior. I’ve never had this level of confusion doing large PRs with BitBucket, switching to GitHub has made it really difficult keeping track of comments across multiple commits in a PR. The current behavior definitely deserves some more finessing, it’s such an important and core capability of GitHub and any git-based tool.</p>

**Comment by MichalLaczynskiAmartus** [comment]
<p>Hi.<br>
What is the status of the issue?<br>
I am also in favour of having all comments in the files changed window.<br>
Now I am forced to llok for other tool than GitHub to deal with my Pull Requests.<br>
And that is such a simple feature to do.<br>
There can always be a flag to trigger “show outdated”</p>
<p>Thanks!</p>

**Comment by codingjeremy** [comment]
<p>+1 to all this. I use gerrit now and it does this so it took me a long time of troubleshooting to figure out the issue wasn’t with me, but the feature isn’t present. Please fix!</p>

**Comment by EsbenVitecsoftware** [comment]
<p>Stupid. This makes me look for alternatives for github. Everyone else knows how to do this.</p>

**Comment by phillips-tech** [comment]
<p>+1 on this.  For a large PR with lots of comments, it is nearly impossible to track.  Can we get a real answer on why this remains an issue?  Does <a href="http://gitlab.com" rel="noopener nofollow ugc">gitlab.com</a> have this issue?</p>

**Comment by kunickiaj** [comment]
<p>Any update on this? Has anyone maybe written a browser extension to fix this? This is the <span class="hashtag">#1</span> and pretty much only reason our org refuses to use PRs for review and we maintain a separate Gerrit instance solely for reviews. Would love to get rid of it if GitHub would fix this.</p>
<p>For those looking for a workaround, I have noticed that the GitHub PR support in Jetbrains IDEs (specifically tested with IntelliJ) provides the view we’re looking for. I suppose it’s better than <em>nothing</em>.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/c/3/c3827cd31ac6f5b2f14ccf6bc69eb197eac137a5.png" data-download-href="/uploads/short-url/rTyFRaJjwaVrC4DPklg5IGPTTp3.png?dl=1" title="Screen Shot 2020-11-10 at 5.25.59 PM" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/746474/181075084-2060c18a-5e0a-40d4-b40a-d24f18834724.png" alt="Screen Shot 2020-11-10 at 5.25.59 PM" data-base62-sha1="rTyFRaJjwaVrC4DPklg5IGPTTp3" width="690" height="113" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/c/3/c3827cd31ac6f5b2f14ccf6bc69eb197eac137a5_2_690x113.png, https://aws1.discourse-cdn.com/github/optimized/3X/c/3/c3827cd31ac6f5b2f14ccf6bc69eb197eac137a5_2_1035x169.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/c/3/c3827cd31ac6f5b2f14ccf6bc69eb197eac137a5_2_1380x226.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/c/3/c3827cd31ac6f5b2f14ccf6bc69eb197eac137a5_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2020-11-10 at 5.25.59 PM</span><span class="informations">3436×566 74.5 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by harishreedharan** [comment]
<p>It would be really useful to add this and migrate full to GitHub. Without this, I can get the buy in to move from our archaic feeling review tool to GitHub. PRs are much better for managing multiple branches and interdependent features for us, but we can’t switch to using PRs unless this is fixed. Thank you!</p>

**Comment by mitchd319** [comment]
@lee-dohm @renozao I am a couple years late to the conversation, but I fully agree it would be much easier for the reviewer if files were updated in the review when new changes are pushed. This is standard in other code review tools, such as Code Collaborator.

I am a relatively inexperienced reviewer in GH, and I was perplexed when I saw that a new commit had been made, but the review did not reflect the changes. In order to resolve comments, I had to cross-reference the review with the specific commit which addressed the changes in another tab. This process was unintuitive and time consuming.

**Reply by mitchd319** [reply]
It looks like I and many others would very much appreciate if @github-staff would either acknowledge the issue or provide a _reasonable_ explanation for why it has not been addressed.

**Comment by ls-jacob-lawrence** [comment]
Tools like [reviewable](https://reviewable.io/) have supported this feature for a really long time.
Everyone I talked to dislikes github reviews compared to other offerings because comments get lost and are difficult to find the context of for reviewers, committers and interested stakeholders.

It would be really helpful if github improved this so we don't have to always use external tools.

**Comment by adorey** [comment]
I agree with all the comments above. It doesn't make sense to not display outdated review comments (or to make it configurable). 

Impossible to make quality reviews on Github which makes you think about other alternatives.

I hope a solution can be given from @github-staff soon.

**Comment by abdulazizab2** [comment]
Will we receive any update regarding this issue in the near future ?

**Comment by precz** [comment]
This issue makes GitHub useless as a review tool. It's how it's done in Gerrit and I miss this feature so much...

**Comment by mathieu-lanefour** [comment]
Currently trialing a switch from our bitbucket instance to github and this issue is a big one, I have PRs with lots of comments and not being able to easily see if the comment was actually addressed is a real problem for us.

Please do something so I can easily see the changes that were made around the comments I have, simply saying outdated isn't enough

**Comment by vinayakakv** [comment]
Gitlab handles this in a Good way by linking to a commit that changed the line in question.

I think we can have something similar to this along with the "outdated" label

**Comment by mlafleur** [comment]
Adding my voice to this as well. The inability to validate the requested change was implemented is extremely frustrating. This issue doesn't exist on any other Git platform I've used, including Azure DevOps.

**Comment by zachovich** [comment]
GitHub naively assumes if a commented line changed, whether to address the comment or for any other reason, that the issue is fixed and the comment "effectively" resolved and no need to be seen anymore! very poor design decision I'd say!

**Comment by rj-wowza** [comment]
we've been complaining about this for years. there are even SASS companies profiting off this exact problem with the github UI. FIX IT ALREADY

**Comment by cssatx** [comment]
I'm somewhat flabbergasted at this. As the lead of a team where we try to make reviewing code as efficient as possible, knowing that a comment has been addressed is of paramount importance. Further, in the name of code review speed, knowing that the ONLY change between one review and the next is code that addresses a comment is EXTREMELY valuable as the reviewer can just do a simple diff. But not in github...

**Comment by szalapski** [comment]
In user story format:

**As a code reviewer, I want to see in a PR what the new code is where my newly resolved comment on old code was, so that I can ensure my concerns were addressed.** (Currently the comment is shown with the old commit but not with the new commit and it is a hassle to figure out what the new code is for a resolved comment.)

> " If the line that the comment was in relation to changed, the automatic tools have no way of knowing if the comment targeted at the old line still makes sense on the new line." Take your best clues and put it somewhere.   When all else fails, put it at the same line number or the end of the file. Something is better than nothing.  A comment on a line that doesn't match the code is better than no comment at all.

No other code review tool gives up when it can't know exactly if the comment is still relevant.  The reviewers have the context of earlier reviews in mind and can work it out, but not if they don't see the comment.

**Comment by gdalam** [comment]
Adding my voice to this. When reviewing a new or junior developer's code I can't just assume that they properly addressed each specific comment, I need to be able to easily see what code changes they made. Even if the review comments were simply linked to specific lines in the file (instead of commit hashes) that would be useful.

**Comment by seangwright** [comment]
Having enjoyed the Azure DevOps and Bitbucket commenting systems for years, I always thought I was doing GitHub PR commenting incorrectly because they wouldnt track when code was updated.

Finding this discussion is both baffling and reassuring. I'm confused why such a strange UX decision would be considered "correct" and happy to find out that it's not me, it's GitHub. 🤦‍♂️

I would love for this "feature" to finally be considered a bug and corrected.

**Comment by that-lucas** [comment]
A few months suffering with this I finally took the time to research about it and I find this discussion.
I am baffled by the absurdity of this "expected behavior".
The person who decided/implemented this clearly doesn't use github to code review his peers' work.

**Comment by Tommuh** [comment]
I review a lot of private code and sometimes I can't remember what I have said earlier. Making it very hard to verify if the made change is the requested change. Especially with 50+ comments for each review.

I do not want to see the old code, I just want to see what my comment was or the comment of my colleague. 
Just keep the comment on the line even if the code is completely gone.

**Comment by gabmis** [comment]
100% agree, please let us see outdated comments in the "files changed" view or make it an opt-in :pray: :pray:

**Comment by ssmulders** [comment]
Adding my vote for this as well. And even if it's intended behaviour, it's not even consistent. One comment will have "outdated" on the Conversation tab when the code is removed. But if the code has changed, you don't see a single thing. Horrible UX...

ETA: the more I look at it, the more I'm convinced this is absolutely a bug. It just doesn't add up. Look at this situation:

<img width="743" alt="image" src="https://github.com/community/community/assets/5196567/5aa0926f-ba2c-43ee-a92f-14fdc0399700">

It says (6) in the top left Conversation tab. So 6 conversations total. Then on the dropdown (and it doesn't change if I change it to "Changes from all commits") it says (4) unresolved conversations and (0) unresolved conversations. That's just wrong - where did the 2 others go? We definitely need at the very least a way to look at the old comments here. 

This was the code change that wasn't even flagged as "outdated" - but it did hide the comment that was attached to the line: 
![image](https://github.com/community/community/assets/5196567/418bf17a-47eb-40e5-b153-602f51c3a39f)

Auto-resolving it? Then why isn't it shown as "resolved"? Whichever way you look at it, it's implemented inconsistently at very best.

And if I switch to watching the files of the commit that the first review was done for, the comment doesn't show up either. This is just a bug.

**Comment by SchroterQuentin** [comment]
There is a solution on Azure DevOps and it's perfect, Github should implement the exact same mechanism. We wouldn't imagine that moving some projet from Azure DevOps to Github would cause us to lose features. Microsoft product owners should discuss together 😄.

**Comment by aw230012** [comment]
GitHub and MIcrosoft are one now right? Just do what Azure DevOps does. I can't believe it's going on 4 years since this was reported and it's still not addressed.

**Comment by stephenjason89** [comment]
@lee-dohm

**Comment by cfbadr** [comment]
This issue is crazy, but what's worse is GitHub's silence about it. This post was created in 2019, and here we are in 2024, still waiting for an answer...

Can you stop mocking your users so openly, GitHub?

**Comment by yuanchen-zhu** [comment]
This is one of the most maddening lack-of-feature I found after switching to GitHub. This makes code reviewing so much more painful compared with Phabricator or gerrit. Come on github devleopers! If Phabricator can implement it with its small team, surely you folks can do it as well. 

I made a new feature requests in the Pull Request section of the board so that hopefully some PM will read it. Please vote there as well if you want to see it implemented (https://github.com/orgs/community/discussions/110796)

**Comment by ew-domain** [comment]
this is ridiculous

**Comment by jxa8057** [comment]
Adding my voice to the choir. It is ridiculous that this issue is being ignored for so long.

**Comment by SultanDino** [comment]
Adding my voice

**Comment by souvik-dey-28** [comment]
hello 1

**Reply by souvik2812** [reply]
hello

**Comment by FirminEx** [comment]
Bump, this is a big UX issue

**Comment by KennethSpeecke** [comment]
Adding my voice to this.
This is the most annoying thing ever not gonna lie.
Not being able to see what has changed on the comment you placed is ridiculous.
Pretty much every other CI Service has this feature.
Please add it...

**Comment by b3nThomas** [comment]
Googled "github make comment not outdated" only to find this thread.
Scratching head why such a simple feature can't be added?
At the very least just have a collapsed comment with the Outdated flag in place in the Files Changed as a marker.
Then you can expand it to see what it was, or click the Outdated pill and a "Remove outdated flag" or something.
I'm refraining from saying things like "This is so dumb" etc as I absolutely love GitHub, this is pretty dumb though. Whoops! 😅

**Comment by jchesshirAspire** [comment]
Hey folks, there's an [Unanswered Post](https://github.com/orgs/community/discussions/17481) out there that's complaining about practically the same thing as this one.  Perhaps extra comments on that one will actually get Github's attention.

**Reply by didmar** [reply]
Here is [another unanswered post](https://github.com/orgs/community/discussions/9956) on the same issue

**Comment by berniedurfee-renaissance** [comment]
Yep, discovered this "feature" today. Makes reviewing really tricky.

**Reply by jchesshirAspire** [reply]
I recommend commenting on one of those other two.  This one might not get seen since the discussion has not been closed.

Admins, I'd like to request further comments to this post be suspended.  It would also be good for one of the above linked posts be selected as the primary one to implement and the other marked as a duplicate.  Which one better describes the problem is up to you.  I think they both do a good job, and both highlight at least one extra problem that the other one doesn't.

**Comment by Tommuh** [comment]
almost 5 years later... unbelievable.

**Comment by IRHayfield** [comment]
I agree that GitHub should change this and make outdated comments an optional feature.
As a workaround I created [outdated-parrot](https://github.com/IRHayfield/outdated-parrot) an action which adds all open outdated comment threads as annotations so they are displayed in the files changed view.
This is not an perfect solution, but does allow users to review comments against the correct line with the addition of some links that users may find useful.

**Comment by zebholt** [comment]
Incredibly annoying PR UI / workflow.

I found this vscode plugin today and it adds the missing feature I was looking for.
Right clicking on the "outdated" comment gives you an option to `Diff comment with HEAD` showing old and new code in one place.

https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-pull-request-github

**Reply by taurus227** [reply]
I can't find this `Diff comment with HEAD` option, could you provide screenshot?

**Comment by Injamul2k24** [comment]
https://github.com/orgs/Hoqueinja/discussions/1

**Reply by jchesshirAspire** [reply]
Not able to follow this link.  Please refresh or remove the comment.

**Comment by Caitlin-Syntax** [comment]
This issue is now documented in at least four places with no real official response.
- https://github.com/orgs/community/discussions/9956
- https://github.com/orgs/community/discussions/17481
- https://github.com/orgs/community/discussions/23138 (this thread)
- https://github.com/orgs/community/discussions/110796 (feature request)

**Comment by CoolGoose** [comment]
I  guess we need to bump all a bit more, it's the most annoying thing coming over gitlab, initial context of seeing how the new change 'fixes it' it's important.

**Comment by serhii-surazhskyi-gl** [comment]
Up

**Comment by dmbursty** [comment]
This 'intended behaviour' doesn't make sense when using Split View, where the left side shows the same file in the exact state the comments were made.  There is no reason the comments can't be attached to the appropriate lines.

**Comment by Caitlin-Syntax** [comment]
New year, new bump. This issue, which has been getting upvotes and comments across at least 4 threads since March 2019, still has no official response. My previous comment links to the threads documenting this issue and requesting UX changes. Please, please fix this, GitHub!

**Comment by elseifthen** [comment]
@lee-dohm I kindly ask you to unmark the "Answer" (or whoever marked it as an answer @renozao?) as it just brushes off what many people request and expect to be natural behavior and functionality (e.g. from gerrit) in order to do efficient code review. This is an issue that needs visibility and absolutely needs to be addressed.

**Reply by Caitlin-Syntax** [reply]
They technically did answer the question (Is this intended? Yes). It's inexplicably bad intended behavior, but it's intended. We might have better luck trying to boost this feature request: https://github.com/orgs/community/discussions/110796

**Comment by Reston** [comment]
devs should go to bitbucket see how they do this thing and copy... because there the comments are kept and you can review properly.

**Reply by layla-retalk** [reply]
This feature is also present in azure devops

**Comment by Caitlin-Syntax** [comment]
Another six months, another new bump. Still awaiting any functional response to any of the (at least) [four discussions](https://github.com/orgs/community/discussions/23138#discussioncomment-10593003) about this issue over six years. This is still making my PR review process profoundly annoying.

**Comment by LionetChen** [comment]
Bump. Azure DevOps has the "View Original Diff", which is handy. It's sorely missed from GitHub.

**Comment by b3nThomas** [comment]
So I just opened `Files Changed` on a PR with the new_files_changed beta active, and my outdated comments are THERE! Granted they're at the top instead of inline, but I almost jumped for joy out of my chair. Huge step in the right direction. Already reaping the benefits on the very first PR with the feature. Finally 🥲

**Reply by kielpins-nearmap** [reply]
I'm trying to find out how to activate the new_files_changed beta, but I'm not able to find it in the repo settings. Google is no help. Could you provide a screenshot? Thanks!

**Reply by stefan-schweiger** [reply]
my assumption would be to add it to the query parameters as `new_files_changed=true` when using the new experience, but not seeing anything different..

**Reply by Tommuh** [reply]
holy moly .. it is added!

**Reply by seangwright** [reply]
What a wonderful day!

![TitanicOldLadyGIF](https://github.com/user-attachments/assets/5d3645cc-7a56-47b7-b010-0a60157d46be)

For anyone unsure how to enable the preview feature, here's a screenshot.

<img width="943" height="518" alt="image" src="https://github.com/user-attachments/assets/b78b7f05-0589-40c5-97f9-159fec1a8e9d" />

**Reply by jwosty** [reply]
GitHub PR review is finally usable? Joyous day!

**Reply by CoolGoose** [reply]
Not yet, since it's just top not inline, but one can keep hope :D

**Reply by serhii-surazhskyi-gl** [reply]
I still see issues with "Outdated" comments not shown even in new feature (at least partially)

**Reply by ardrabczyk** [reply]
Where can you see it @b3nThomas? Outdated comments are listed under Conversations but only in the old view after pressing 'Preview Switch back' near the upper right corner and the 'New Files Changed Experience' doesn't affect it. For example at https://github.com/lee-dohm/test-repo/pull/6/files:

<img width="2512" height="674" alt="HERE" src="https://github.com/user-attachments/assets/e8f78a30-a551-43ca-b09c-82d87070ca11" />

**Reply by ardrabczyk** [reply]
After pressing 'Try the new experience' which I think is an equivalent of enabling New Files Changed Experience in 'Feature preview' there is a comment icon next to 'Submit review' button but comments started by me are not listed here, also filtering doesn't work, is the same experience you get @serhii-surazhskyi-gl @b3nThomas?

**Reply by serhii-surazhskyi-gl** [reply]
Yes, but I'm usually in commit view (even it moves to `Files Changed`).
I tested this functionality for last few months - it is garbage with the same problems.
There are a lot of outdated comments that are not shown, even they are not resolved.
In addition there are resolved comments shown in review, even they are resolved...

**Comment by quackemi-inmar** [comment]
Still fighting this in 2026.  Makes the PR process nearly unusable.

**Comment by stephen-trfg** [comment]
+1 in 2026. Coming from DevOps, where this is handled well, it’s surprising to see such a limiting design choice.

**Comment by dafi-dnv** [comment]
Azure DevOps review works so well. Comments are almost always where they were added. In very rare cases it's on the file, but it is still visible. GitHub PR review is huge pain, because the whole point of leaving comment in PR is to change the code, and then resolve if current code resolves the issue. 🤯

**Comment by Emprint** [comment]
Our project switched from DevOps to Github and this is so far the most painfully missing feature on github, the ability to quickly see how a comment was solved by seeing the new file changes next to the previous version with the comment.

**Comment by Yaroslav-Bershadskyi-P1** [comment]
bump!

**Comment by wlitheng** [comment]
Azure DevOps handles review comments much better. Comments usually stay attached to the correct code, and even if they don't, they're still visible. 

GitHub PR reviews are frustrating because comments often disappear from the diff after code changes, making them hard to follow up on.

I have to open 2 tabs.
1 tab to view the resolved comments and to know which file, which line.
1 tab to view the exact code line to know what change was made to resolve the comment.
Utterly ridiculous! Rubbish!

**Reply by Ju6475** [reply]
YuriyUniyat

**Comment by hari90** [comment]
+1 This is a huge setback for our team coming especially given we are moving over from an older system that handled it better.

**Comment by Flaming-Dorito** [comment]
It's been more than 7 years since this thread was started. I think we just have to accept that GitHub doesn't care. After they labeled this "Intentional" they won't reconsider whether it's a good idea.

Literally EVERY single version control website handles this issue better than GitHub, but they simply do not care, so we are just yelling into the void.

Look at the response to the "Official" accepted Answer that says this is intentional behavior. Apparently if something is intentional it can't be a bad idea.
<img width="860" height="159" alt="image" src="https://github.com/user-attachments/assets/cea4a1b2-bd29-4403-aa90-be51166264fa" />

They KNOW no users like this behavior. They DON'T CARE.

The only solution is to move towards migrating to better services than GitHub.

**Comment by Caitlin-Syntax** [comment]
Given that this thread seems to be wholly ignored by GitHub because it's marked as "answered" (lol), I'd recommend we all go boost the open feature request listed below. This issue has been documented in (at least) the four threads below since 2019. I wonder if we're going to make it to a decade with no fix...

- https://github.com/orgs/community/discussions/9956
- https://github.com/orgs/community/discussions/17481
- https://github.com/orgs/community/discussions/23138 (this thread)
- https://github.com/orgs/community/discussions/110796 (feature request)
